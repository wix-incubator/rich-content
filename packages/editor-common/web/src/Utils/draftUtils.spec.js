import mockLinkEditorState from '../../../../../e2e/tests/fixtures/headers.json';
import mockAlignmentEditorState from '../../../../../e2e/tests/fixtures/text-alignment.json';
import mockGiphyEditorState from '../../../../../e2e/tests/fixtures/giphy.json';

import { EditorState, convertToRaw, convertFromRaw } from '@wix/draft-js';
import {
  insertLinkAtCurrentSelection,
  insertLinkInPosition,
  hasLinksInBlock,
  hasLinksInSelection,
  getLinkDataInSelection,
  removeLinksInSelection,
  setTextAlignment,
  getAnchorBlockData,
  isAtomicBlockFocused,
  replaceWithEmptyBlock,
  deleteBlock,
  getSelectedBlocks,
  getTextAlignment,
  mergeBlockData,
  setEntityData,
  createBlock,
  getSelectionRange,
  isInSelectionRange,
  createEntity,
  getFocusedBlockKey,
  getBlockInfo,
  setSelection,
} from './draftUtils';

const getContentAsObject = editorState => convertToRaw(editorState.getCurrentContent());
const getEditorStateFromJson = json => EditorState.createWithContent(convertFromRaw(json));
const setEditorStateSelection = (editorState, selection) => {
  const newSelection = editorState.getSelection().merge(selection);
  return EditorState.forceSelection(editorState, newSelection);
};
const getEditorStateWithSelectionAt = (editorState, blockKey) => {
  return setEditorStateSelection(editorState, {
    anchorKey: blockKey,
    anchorOffset: 0,
    focusKey: blockKey,
    focusOffset: 1,
  });
};
const linkData = {
  url: 'www.google.com',
  targetBlank: true,
  nofollow: false,
  anchorTarget: '_blank',
  relValue: 'noopener',
};
const BLOCK_KEY1 = '50k2j';
const BLOCK_KEY2 = '1u5r4';
const BLOCK_KEY3 = 'd4hhc';
const BLOCK_KEY4 = '8s1v3';
const BLOCK_KEY5 = 'bsrvp';

const selection1 = {
  anchorKey: BLOCK_KEY1,
  anchorOffset: 0,
  focusKey: BLOCK_KEY1,
  focusOffset: 7,
};

const selection2 = {
  anchorKey: BLOCK_KEY2,
  anchorOffset: 0,
  focusKey: BLOCK_KEY5,
  focusOffset: 0,
};

const selection3 = {
  anchorKey: BLOCK_KEY2,
  anchorOffset: 0,
  focusKey: BLOCK_KEY2,
  focusOffset: 0,
};

const htmlData = {
  srcType: 'html',
  src: '<p> Hello World! </p>',
  config: { alignment: 'center', width: '350' },
};

const gifData = {
  config: {
    size: 'content',
    alignment: 'center',
  },
  gif: {
    originalUrl: 'https://media1.giphy.com/media/hrk8ehR4lCZ27FtjPA/giphy_s.gif',
    stillUrl: 'https://media1.giphy.com/media/hrk8ehR4lCZ27FtjPA/giphy_s.gif',
    height: 400,
    width: 480,
  },
};

const htmlType = 'wix-draft-plugin-html';

describe('Test draftUtils functions', () => {
  describe('Test draftUtils Links functions', () => {
    const editorState = getEditorStateFromJson(mockLinkEditorState);

    it('Test insertLinkInPosition function', () => {
      const editorStateWithLink = insertLinkInPosition(editorState, BLOCK_KEY1, 0, 7, linkData);
      const contentStateObj = getContentAsObject(editorStateWithLink);
      expect(contentStateObj).toMatchSnapshot();
    });

    it('Test insertLinkAtCurrentSelection function', () => {
      const editorStateWithSelection = setEditorStateSelection(editorState, selection1);
      const editorStateWithLink = insertLinkAtCurrentSelection(editorStateWithSelection, linkData);
      const contentStateObj = getContentAsObject(editorStateWithLink);
      expect(contentStateObj).toMatchSnapshot();
    });

    describe('Test hasLinksInBlock function', () => {
      const editorStateWithLink = insertLinkInPosition(editorState, BLOCK_KEY1, 0, 7, linkData);
      const contentState = editorStateWithLink.getCurrentContent();
      it('should return true on block with link', () => {
        expect(hasLinksInBlock(contentState.getBlockForKey(BLOCK_KEY1), contentState)).toEqual(
          true
        );
      });
      it('should return false on block without link', () => {
        expect(hasLinksInBlock(contentState.getBlockForKey(BLOCK_KEY3), contentState)).toEqual(
          false
        );
      });
    });

    const selection = editorState.getSelection().merge(selection1);
    const editorStateWithLink = insertLinkInPosition(editorState, BLOCK_KEY1, 0, 7, linkData);
    const editorStateWithSelectionOnLink = setEditorStateSelection(editorStateWithLink, selection);

    describe('Test hasLinksInSelection function', () => {
      const editorStateWithoutLinks = EditorState.forceSelection(editorState, selection);

      it('should return true on editor state with link in selection', () => {
        expect(hasLinksInSelection(editorStateWithSelectionOnLink)).toEqual(true);
      });
      it('should return false on editor state without links in selection', () => {
        expect(hasLinksInSelection(editorStateWithoutLinks)).toEqual(false);
      });
      it('should return false on editor state without selections at all', () => {
        expect(hasLinksInSelection(editorState)).toEqual(false);
      });
    });

    it('Test getLinkDataInSelection function', () => {
      const selectionLinkData = getLinkDataInSelection(editorStateWithSelectionOnLink);
      expect(selectionLinkData).toMatchSnapshot();
    });

    describe('Test removeLinksInSelection function', () => {
      it('should remove link in selection', () => {
        const newEditorState = removeLinksInSelection(editorStateWithSelectionOnLink);
        const contentStateObj = getContentAsObject(newEditorState);
        expect(contentStateObj).toMatchSnapshot();
      });
      it('should not remove links outside the selection', () => {
        const newEditorState = removeLinksInSelection(editorStateWithLink);
        const contentStateObj = getContentAsObject(newEditorState);
        expect(contentStateObj).toMatchSnapshot();
      });
    });
  });

  describe('Test draftUtils Alignment functions', () => {
    const editorState = getEditorStateFromJson(mockAlignmentEditorState);
    const editorStateWithSelection = getEditorStateWithSelectionAt(editorState, BLOCK_KEY1);
    it('Test getTextAlignment function', () => {
      expect(getTextAlignment(editorState)).toEqual('left');
    });
    it('Test setTextAlignment function', () => {
      const newEditorState = setTextAlignment(editorStateWithSelection, 'right');
      const contentStateObj = getContentAsObject(newEditorState);
      expect(contentStateObj).toMatchSnapshot();
    });
    it('Test getAnchorBlockData function', () => {
      const anchorBlockData = getAnchorBlockData(editorStateWithSelection);
      expect(anchorBlockData).toMatchSnapshot();
    });
  });

  describe('Test draftUtils Blocks functions', () => {
    const editorState = getEditorStateFromJson(mockGiphyEditorState);
    const editorStateWithSelectionOnAtomic = getEditorStateWithSelectionAt(editorState, BLOCK_KEY2);
    const editorStateWithSelectionOnNotAtomic = getEditorStateWithSelectionAt(
      editorState,
      BLOCK_KEY4
    );

    it('Test replaceWithEmptyBlock function', () => {
      const newEditorState = replaceWithEmptyBlock(editorStateWithSelectionOnAtomic, BLOCK_KEY2);
      const contentStateObj = getContentAsObject(newEditorState);
      expect(contentStateObj).toMatchSnapshot();
    });

    it('Test deleteBlock function', () => {
      const newEditorState = deleteBlock(editorState, BLOCK_KEY2);
      const contentStateObj = getContentAsObject(newEditorState);
      expect(contentStateObj).toMatchSnapshot();
    });

    describe('Test isAtomicBlockFocused function', () => {
      it('should return false for no focus on atomic block', () => {
        expect(isAtomicBlockFocused(editorStateWithSelectionOnNotAtomic)).toEqual(false);
      });
      it('should return true for focus on atomic block', () => {
        expect(isAtomicBlockFocused(editorStateWithSelectionOnAtomic)).toEqual(true);
      });
    });

    describe('Test getSelectedBlocks function', () => {
      const editorWithSelectedBlocks = setEditorStateSelection(editorState, selection2);

      it('should return the block that belongs to the cursor', () => {
        const selectedBlocks = getSelectedBlocks(editorState);
        expect(selectedBlocks).toMatchSnapshot();
      });
      it('should return array with the selected blocks', () => {
        const selectedBlocks = getSelectedBlocks(editorWithSelectedBlocks);
        expect(selectedBlocks).toMatchSnapshot();
      });
    });

    it('should merge block data', () => {
      const editorState = getEditorStateFromJson(mockAlignmentEditorState);
      const editorStateWithSelection = setEditorStateSelection(editorState, selection1);
      const mergedBlockData = mergeBlockData(editorStateWithSelection, { textAlignment: 'right' });
      expect(mergedBlockData).toMatchSnapshot();
    });

    it('should set entity data', () => {
      const editorState = getEditorStateFromJson(mockGiphyEditorState);
      const withNewEntity = setEntityData(editorState, 0, { alignment: 'center', size: 'small' });
      expect(withNewEntity).toMatchSnapshot();
    });

    it('should create block', () => {
      const editorState = getEditorStateFromJson(mockGiphyEditorState);
      const { newBlock } = createBlock(editorState, htmlData, htmlType);
      const blockData = newBlock.data;
      expect(blockData).toMatchSnapshot();
    });

    it('should get selection renge', () => {
      const editorState = getEditorStateFromJson(mockAlignmentEditorState);
      const editorStateWithSelection = setEditorStateSelection(editorState, selection1);
      const selection = getSelectionRange(
        editorStateWithSelection,
        editorStateWithSelection.getCurrentContent().getBlockForKey(BLOCK_KEY1)
      );
      expect(selection).toMatchSnapshot();
    });

    it('should return if range includes in selection', () => {
      expect(isInSelectionRange([3, 7], [0, 9])).toEqual(true);
      expect(isInSelectionRange([3, 7], [8, 9])).toEqual(false);
    });

    it('should create entity and return the new entity key', () => {
      const editorState = getEditorStateFromJson(mockGiphyEditorState);
      expect(createEntity(editorState, { type: htmlType, data: htmlData })).toEqual('10');
    });

    it('should get focused block key', () => {
      const editorState = getEditorStateFromJson(mockGiphyEditorState);
      const editorWithSelectedBlocks = setEditorStateSelection(editorState, selection3);
      expect(getFocusedBlockKey(editorWithSelectedBlocks)).toEqual(BLOCK_KEY2);
    });

    it('should get block info', () => {
      const editorState = getEditorStateFromJson(mockGiphyEditorState);
      const blockInfo = getBlockInfo(editorState, BLOCK_KEY2);
      expect(blockInfo).toMatchSnapshot();
    });

    it('should set selection', () => {
      const editorState = getEditorStateFromJson(mockGiphyEditorState);
      const editorWithSelection = setSelection(editorState, selection2);
      expect(editorWithSelection).toMatchSnapshot();
    });
  });
});
