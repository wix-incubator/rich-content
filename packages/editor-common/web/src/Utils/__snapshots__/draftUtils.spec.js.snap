// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test draftUtils functions Test draftUtils Alignment functions Test getAnchorBlockData function 1`] = `
Object {
  "textAlignment": "left",
}
`;

exports[`Test draftUtils functions Test draftUtils Alignment functions Test setTextAlignment function 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {
        "textAlignment": "right",
      },
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "50k2j",
      "text": "I am text without alignment",
      "type": "unstyled",
    },
    Object {
      "data": Object {
        "textAlignment": "left",
      },
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "civd5",
      "text": "I am text with left alignment",
      "type": "unstyled",
    },
    Object {
      "data": Object {
        "textAlignment": "center",
      },
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "ehr1q",
      "text": "I am text with center alignment",
      "type": "unstyled",
    },
    Object {
      "data": Object {
        "textAlignment": "right",
      },
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "bmg0d",
      "text": "I am text with right alignment",
      "type": "unstyled",
    },
    Object {
      "data": Object {
        "textAlignment": "justify",
      },
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "e2imv",
      "text": "I am text with justify alignment",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {},
}
`;

exports[`Test draftUtils functions Test draftUtils Blocks functions Test deleteBlock function 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "8s1v3",
      "text": "",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "bsrvp",
      "text": "",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {},
}
`;

exports[`Test draftUtils functions Test draftUtils Blocks functions Test getSelectedBlocks function should return array with the selected blocks 1`] = `
Array [
  Immutable.Record {
    "key": "1u5r4",
    "type": "atomic",
    "text": " ",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": "3",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  Immutable.Record {
    "key": "bsrvp",
    "type": "unstyled",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
]
`;

exports[`Test draftUtils functions Test draftUtils Blocks functions Test getSelectedBlocks function should return the block that belongs to the cursor 1`] = `
Array [
  Immutable.Record {
    "key": "8s1v3",
    "type": "unstyled",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
]
`;

exports[`Test draftUtils functions Test draftUtils Blocks functions Test replaceWithEmptyBlock function 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "8s1v3",
      "text": "",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "1u5r4",
      "text": "",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "bsrvp",
      "text": "",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {},
}
`;

exports[`Test draftUtils functions Test draftUtils Links functions Test getLinkDataInSelection function 1`] = `
Object {
  "rel": "noopener",
  "target": "_blank",
  "url": "www.google.com",
}
`;

exports[`Test draftUtils functions Test draftUtils Links functions Test insertLinkAtCurrentSelection function 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 0,
          "length": 7,
          "offset": 0,
        },
      ],
      "inlineStyleRanges": Array [
        Object {
          "length": 7,
          "offset": 0,
          "style": "UNDERLINE",
        },
      ],
      "key": "50k2j",
      "text": "Leverage agile frameworks to provide a robust synopsis for high level overviews. Iterative approaches to corporate strategy foster collaborative thinking to further the overall value proposition.",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "d4hhc",
      "text": "Bring to the table win-win survival strategies to ensure proactive domination.",
      "type": "header-two",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "2s6ta",
      "text": "At the end of the day, going forward, a new normal that has evolved from generation X is on the runway heading towards a streamlined cloud solution.",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "87u",
      "text": "Organically grow the holistic world view of disruptive innovation via workplace diversity and empowerment.",
      "type": "header-three",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "am5bt",
      "text": "Capitalize on low hanging fruit to identify a ballpark value added activity to beta test.",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {
    "0": Object {
      "data": Object {
        "rel": "noopener",
        "target": "_blank",
        "url": "www.google.com",
      },
      "mutability": "MUTABLE",
      "type": "LINK",
    },
  },
}
`;

exports[`Test draftUtils functions Test draftUtils Links functions Test insertLinkInPosition function 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 0,
          "length": 7,
          "offset": 0,
        },
      ],
      "inlineStyleRanges": Array [
        Object {
          "length": 7,
          "offset": 0,
          "style": "UNDERLINE",
        },
      ],
      "key": "50k2j",
      "text": "Leverage agile frameworks to provide a robust synopsis for high level overviews. Iterative approaches to corporate strategy foster collaborative thinking to further the overall value proposition.",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "d4hhc",
      "text": "Bring to the table win-win survival strategies to ensure proactive domination.",
      "type": "header-two",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "2s6ta",
      "text": "At the end of the day, going forward, a new normal that has evolved from generation X is on the runway heading towards a streamlined cloud solution.",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "87u",
      "text": "Organically grow the holistic world view of disruptive innovation via workplace diversity and empowerment.",
      "type": "header-three",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "am5bt",
      "text": "Capitalize on low hanging fruit to identify a ballpark value added activity to beta test.",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {
    "0": Object {
      "data": Object {
        "rel": "noopener",
        "target": "_blank",
        "url": "www.google.com",
      },
      "mutability": "MUTABLE",
      "type": "LINK",
    },
  },
}
`;

exports[`Test draftUtils functions Test draftUtils Links functions Test removeLinksInSelection function should not remove links outside the selection 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 0,
          "length": 7,
          "offset": 0,
        },
      ],
      "inlineStyleRanges": Array [
        Object {
          "length": 7,
          "offset": 0,
          "style": "UNDERLINE",
        },
      ],
      "key": "50k2j",
      "text": "Leverage agile frameworks to provide a robust synopsis for high level overviews. Iterative approaches to corporate strategy foster collaborative thinking to further the overall value proposition.",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "d4hhc",
      "text": "Bring to the table win-win survival strategies to ensure proactive domination.",
      "type": "header-two",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "2s6ta",
      "text": "At the end of the day, going forward, a new normal that has evolved from generation X is on the runway heading towards a streamlined cloud solution.",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "87u",
      "text": "Organically grow the holistic world view of disruptive innovation via workplace diversity and empowerment.",
      "type": "header-three",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "am5bt",
      "text": "Capitalize on low hanging fruit to identify a ballpark value added activity to beta test.",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {
    "0": Object {
      "data": Object {
        "rel": "noopener",
        "target": "_blank",
        "url": "www.google.com",
      },
      "mutability": "MUTABLE",
      "type": "LINK",
    },
  },
}
`;

exports[`Test draftUtils functions Test draftUtils Links functions Test removeLinksInSelection function should remove link in selection 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "50k2j",
      "text": "Leverage agile frameworks to provide a robust synopsis for high level overviews. Iterative approaches to corporate strategy foster collaborative thinking to further the overall value proposition.",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "d4hhc",
      "text": "Bring to the table win-win survival strategies to ensure proactive domination.",
      "type": "header-two",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "2s6ta",
      "text": "At the end of the day, going forward, a new normal that has evolved from generation X is on the runway heading towards a streamlined cloud solution.",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "87u",
      "text": "Organically grow the holistic world view of disruptive innovation via workplace diversity and empowerment.",
      "type": "header-three",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "am5bt",
      "text": "Capitalize on low hanging fruit to identify a ballpark value added activity to beta test.",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {},
}
`;
